Vectorization + minor fixes

Sholde opened the pull request.

Sholde committed:
test instrumentation of vector arithmetic operation

- check if wrapper functions are correctly called
- support of each operation : add, mul, sub and div
- support of each size      : 2, 4, 8, 16

Sholde committed:
add test to test result of arithmetic operation

- basic test
- support all size      : 2, 4, 8, 16
- support all operation : add, mul, sub, div

Sholde committed:
Merge pull request #2 from Sholde/test-vector

add test to test result of arithmetic operation

Sholde committed:
add vector function in interflop interface

- init with NULL pointer for the moment
- all backends support

Sholde committed:
change function name

Sholde committed:
test if vector instruction are used

- check architecture and test for x86_64
- check instruction set
- display instruction set for the moment

Sholde committed:
basic test for instumentation set of x86_64 architecture

- test for addition only
- display if test pass for function instrumentation (wrapper call) and insrumentation set test

Sholde committed:
check for all x86_64 instructtion set instruction and register used

Sholde committed:
add const keyword to avoid warning

Sholde committed:
implement ieee vector interface function

Sholde committed:
factorize vector wrapper with size parameter and call vector backend interface

Sholde committed:
change name of macro which defined arithmetic vector wrapper

Sholde committed:
add probes call vector cmp backend functions

Sholde committed:
add checking if one backend support vector operation

- check on vector wrapper (not optimal, in preference check on vfc_init)

Sholde committed:
add implementation of vprec vector interface in scalar mode

Sholde committed:
fix formating

Sholde committed:
add test for ieee, vprec and mca backends

Sholde committed:
keep log of wrapper instrumentation

- create wrapper_log folder and put all .ll into

Sholde committed:
remove opti flags and add save-temps log

- optimization flags call scalar wrapper because it is the same compute for all aperands
- add log but not check automaticly

Sholde committed:
merge instrumentation and result of operation of vector test

- remove mca backend because vector function are not implemented
- fix size of vector test to 2 because other don't work with this condition

Sholde committed:
add in comment mca backend and other size of vector

Sholde committed:
remove instruction set test and enhance print result of test

Sholde committed:
Merge pull request #3 from Sholde/test-vector

basic test for vector operation instrumentation

Sholde committed:
Merge pull request #4 from Safecarlo/test-vector

basic test for vector operation instrumentation

Sholde committed:
Merge pull request #5 from Safecarlo/vector-interface

vector interface for ieee and vprec backend on scalar mode

Sholde committed:
add test for mca backend

Sholde committed:
add another result file to just test with vector of size 2

Sholde committed:
change vector test because mca set 0 to -0 sometimes

Sholde committed:
implementation of vector operation in mca backend in scalar mode

Sholde committed:
cut assembler of backend into vector function file

Sholde committed:
create bash function to check vector instruction and register

Sholde committed:
add test to check if vector instruction and register are used in vector backend

Sholde committed:
Merge pull request #6 from Sholde/vectorization

test and implementation for vector in mca backend in scalar mode

Sholde committed:
Merge pull request #7 from Sholde/test-vector-register-instruction

test vector register and instruction

Sholde committed:
enhance test display

Sholde committed:
Merge pull request #8 from Sholde/test-vector

enhance test display

Sholde committed:
fix precision problem

Sholde committed:
move definition of vector type

- vector backends need it

Sholde committed:
vectorization of ieee backend

- enable -march=native cflags to allow clang to vectorize the backend
  - use a macro to define the vector operation
  - need to test the size because we need to cast in vector type

Sholde committed:
Merge pull request #11 from Sholde/vector-interface

vectorization of ieee backend

Sholde committed:
move function check vector in init wrapper function

Sholde committed:
vector init for c to NAN

Sholde committed:
remove const vector input on interface and needed backend

Sholde committed:
Merge pull request #12 from Sholde/vector-interface

some update on vector interface

Sholde committed:
enhance test

- generate result for our cpu architecture
- stock all binary
- check result only for our cpu architecture vector
- add list in begining of the test whoch contain test parameter
- add more comment

Sholde committed:
formatting

Sholde committed:
Merge branch 'vectorization' into test-vector

Sholde committed:
Merge pull request #13 from Sholde/test-vector

enhance vector test

Sholde committed:
fix format

Sholde committed:
formatting

Sholde committed:
Merge pull request #14 from Sholde/test-vector

formatting vector test

Sholde committed:
check result after run all test (precision, size)

Sholde committed:
Merge pull request #15 from Sholde/test-vector

check result after run all test (precision, size)

Sholde committed:
enable vector operation for ieee, vprec and mca backends

- add -march=native to enable automaticaly the higher vector size
- compile wrapper with -march=native
- compile cited backends with -march=native

Sholde committed:
Merge pull request #16 from Sholde/vector-interface

enable vector operation for ieee, vprec and mca backends

Sholde committed:
remove clang warning for const cast in ieee

Sholde committed:
Merge pull request #17 from Sholde/vector-interface

remove clang warning for const cast in ieee

Sholde committed:
vectorize operation in vprec backend

Sholde committed:
Merge pull request #18 from Sholde/vector-interface

vectorize operation in vprec backend

Sholde committed:
fix formatting

Sholde committed:
Merge pull request #19 from Sholde/vector-interface

fix formatting

Sholde committed:
work in progress

Sholde committed:
add vector function in scalar mode for other backends

Sholde committed:
add vector test for bitmask and mca_mprf

- don't add cancellation because it need specific check (much cancellation detected -> change result)

Sholde committed:
Merge pull request #20 from Sholde/test-vector

add vector test for bitmask and mca_mprf

Sholde committed:
Merge pull request #21 from Sholde/vector-interface

add vector function in scalar mode for other backends

Sholde committed:
add cancellation backend test

Sholde committed:
Merge pull request #22 from Sholde/test-vector

add cancellation backend test

Sholde committed:
better comment

Sholde committed:
formating

Sholde committed:
Merge pull request #23 from Sholde/vector-interface

Vector interface

Sholde committed:
add vector test for ieee test

- just test with double2 which is SSE (because lot of machine have SSE and sometimes not AVX)

Sholde committed:
fix ir test

Sholde committed:
fix type test problem

Sholde committed:
fix size of vector

Sholde committed:
add test for vector type on debug mode

- also clear directory after test

Sholde committed:
Merge pull request #26 from Sholde/test-vector

add specific test for vector type in ieee backend

Sholde committed:
formatting

Sholde committed:
Merge pull request #27 from Sholde/vector-interface

formatting

Sholde committed:
add vector test for vprec backend

Sholde committed:
Merge pull request #28 from Sholde/test-vector

add vector test for vprec backend

Sholde committed:
Merge branch 'master' into vectorization

Sholde committed:
Merge pull request #30 from Sholde/vectorization

update from master of verificarlo

Sholde committed:
Merge branch 'vectorization' into vectorize-vprec

Sholde committed:
work in progress

Sholde committed:
work in progress

Sholde committed:
add vector macro for rounding 32 but it is scalar

Sholde committed:
work in progress

Sholde committed:
Merge pull request #31 from Sholde/vectorize-vprec

add vector macro for rounding 32 but it is scalar

Sholde committed:
work in progress

Sholde committed:
vectorize the rounding of binary32 normal

Sholde committed:
Merge pull request #32 from Sholde/vectorize-vprec

vectorize the rounding of binary32 normal

Sholde committed:
work in progress

Sholde committed:
add normal vector test in vprec

Sholde committed:
Merge branch 'vectorize-vprec' of github.com:Sholde/verificarlo into vectorize-vprec

Sholde committed:
adding compute absErr vprec binary32 vector function

Sholde committed:
work in progress

Sholde committed:
work in progress

Sholde committed:
remove bit-field with vector type

Sholde committed:
macro to simulate bit-field for vector type

Sholde committed:
change c compiler from gcc to clang, to enable vectorization

Sholde committed:
Merge pull request #33 from Sholde/vectorize-vprec

adding vectorisation of vprec for single precision

Sholde committed:
Merge master

Sholde committed:
adding new int32 vector type to respect scalar type

Sholde committed:
create a separate to define vector type

Sholde committed:
rename vector function

Sholde committed:
update clean.sh with vector files

Sholde committed:
factorize size vector check

Sholde committed:
create double vector structure for vprec

Sholde committed:
Merge pull request #34 from Sholde/vectorize-vprec

clean version for float vector

Sholde committed:
adding vector operations count

Sholde committed:
Merge pull request #35 from Sholde/ieee-vector-count

adding vector operations count

Sholde committed:
change type in macro to set bit field on vector

Sholde committed:
addind double vector tests

Sholde committed:
Merge branch 'vectorize-vprec' of github.com:Sholde/verificarlo into vectorize-vprec

Sholde committed:
Merge pull request #36 from Sholde/vectorize-vprec

update

Sholde committed:
improve count op output

Sholde committed:
adding support of double vector in vprec in vector mode

Sholde committed:
adding specific pourcentage of vector size

Sholde committed:
change output format of operaions count

Sholde committed:
using one loop to treat the vector

Sholde committed:
uniform test between float and double, and separate test with echo

Sholde committed:
fix comment line size

Sholde committed:
Merge pull request #37 from Sholde/vectorization

some update - stable

Sholde committed:
Merge branch 'master' into vectorization

Sholde committed:
Merge pull request #39 from Sholde/vectorization

update with master branch

Sholde committed:
fix precision problem in vprec test

Sholde committed:
Merge pull request #40 from Sholde/vectorization

fix precision problem in vprec test

Sholde committed:
adding else if to avoid useless if test in the loop

Sholde committed:
Merge pull request #41 from Sholde/vectorization

adding else if to avoid useless if test in the loop

Sholde committed:
change vector interface

- 1 function by size
- just IEEE and VPREC are implemented
- need to implement others backend with the new interface
- fix vprec format for vector normal

Sholde committed:
fix comment

Sholde committed:
implement the new vector interface for bitmask backend

Sholde committed:
implement cancellation backend with new interface

Sholde committed:
re-indent

Sholde committed:
implement mca backend with new interface

Sholde committed:
fix macro for bitmask, work for double now

Sholde committed:
implement mca-mpfr backend with new interface

Sholde committed:
fix vector function for ieee

Sholde committed:
increase speedup of ieee by removing test of size to count op and one test to debug

Sholde committed:
remove test to count op, increment directly

Sholde committed:
enable test of count op options

Sholde committed:
update clean file

Sholde committed:
factorisze test

Sholde committed:
fix type and remove useless test of set[i]

pablooliveira committed:
Reorganize documentation into chapters

pablooliveira committed:
Fix formatting

pablooliveira committed:
Fix markdown summary links

pablooliveira committed:
Rework "How to cite" section to include more publications

pablooliveira committed:
Update copyright year

pablooliveira committed:
Reorganize documentation and split DD and inclusion/exclusion

pablooliveira committed:
Add list and links to backends in the relevant chapter

Sholde committed:
using float_type.h header in interflop.h

Sholde committed:
adding float_type.h to include path

Sholde committed:
remove superfluous functions in vprec

Sholde committed:
inlining all vector macros in vprec and simplifying of perform vector op

Sholde committed:
fix overridding

Sholde committed:
separate file for vector functions of ieee

Sholde committed:
separate file for vector functions in vprec

Sholde committed:
fix typo hasmap -> hashmap

Sholde committed:
change error to warning

Sholde committed:
remove vector instrumentation tests

titeup commented:
Thanks for the PR, We will shift our discution here.
As far as I remember our last meeting,  regarding the support of  backends not implementing the vector interface, the idea is have null value in the interface, in the wrapper, test if the backend implement this feature by looking for these null values. if it is implemented for the target size forward the call to the backend, if not, serialize.

Sholde committed:
unpack if backend doesn't support vectorization

Sholde committed:
remove vector interface from bitmask

Sholde committed:
remove vector interface from mca

Sholde committed:
using NULL instead of 0 for null pointer

Sholde commented:
Normally I did that.

Sholde commented on src/backends/interflop-cancellation/interflop_cancellation.c:
```diff
@@ -106,6 +106,27 @@ static inline double _noise_binary64(const int exp) {
     }                                                                          \
   })
 
+#define define_interflop_add_sub_vector(size, precision, ops, ope)             \
```
For the backend cancellation, would you keep that ?
For the moment the backend is not really vectored because I didn't vectorize `cancell` function.
But multiplication and division didn't use this function, then they are vectored.

Sholde committed:
remove vector interface from mca-mpfr

Sholde commented on src/common/float_type.h:
```diff
@@ -0,0 +1,43 @@
+#define __FLOAT_TYPE_H_
+
+#include <stdint.h>
+
```
I group together all vector types because I need them in a lot of file like specific file fro `vprec` (`vprec_tools` and `float_struct`) and I think it is not good to include the file `interflop.h` just to have these vector type, also 2 separated files for usual and specialized types.

Sholde committed:
fix unpack version

Sholde commented on src/backends/interflop-vprec/Makefile.am:
```diff
@@ -4,6 +4,7 @@ libinterflop_vprec_la_CFLAGS = -DBACKEND_HEADER="interflop_vprec"
 if WALL_CFLAGS
 libinterflop_vprec_la_CFLAGS += -Wall -Wextra -Wno-varargs
 endif
+libinterflop_vprec_la_CFLAGS += -march=native
```
I need the flag `-march=native` for all backends which use vector interface else it will not work because the compiler can't match vector types with expected vector set instruction.

Sholde commented on verificarlo.in.in:
```diff
@@ -131,7 +131,7 @@ def linker_mode(sources, options, output, args):
     extra_args += "-DINST_FUNC " if args.inst_func else ""
 
-    shell('{clang} -c -O3 -Wno-varargs {extra_args} -o .vfcwrapper.o {vfcwrapper} -I {mcalib_includes}'.format(
+    shell('{clang} -c -O3 -march=native -Wno-varargs {extra_args} -o .vfcwrapper.o {vfcwrapper} -I {mcalib_includes}'.format(
```
It is just here to patch partially vectors problem, but it may be remove after yohan PR.

Sholde committed:
fix typos

Sholde commented on src/backends/interflop-ieee/interflop_ieee.c:
```diff
@@ -327,6 +335,9 @@ static void _interflop_cmp_double(const enum FCMP_PREDICATE p, const double a,
 
+// Include vector operations
+#include "interflop_ieee_vector.h"
+
```
In vector files I use macro that are defined on the original file.
It is why the include is in the middle.

Sholde replied:
I think I can put this `.h` file in a `.c` file instead and it will fix the problem.

Sholde replied:
Actually, I don't know if it will fix the problem because we will still need other macro. I don't know.

Sholde commented on src/backends/interflop-vprec/interflop_vprec.c:
```diff
@@ -1205,6 +1207,12 @@ void _interflop_exit_function(interflop_function_stack_t *stack, void *context,
+ ***************************************************************/
+
+#include "interflop_vprec_vector.h"
+
```
For vprec too.

Sholde committed:
remove extra space

Sholde commented on src/backends/interflop-vprec/interflop_vprec.c:
```diff
@@ -1549,11 +1557,51 @@ struct interflop_backend_interface_t interflop_init(int argc, char **argv,
       _interflop_mul_float,
       _interflop_div_float,
       NULL,
+      _interflop_add_float_2x,
```
I have a warning at each functions which use vectors types beyond the max vector width supported.
But for vprec there is also vprec_tools which use it then I don't know how to remove the warning.
Should I put a diagnostic to avoid at each vector functions ?

Sholde replied:
As I added a file only for vector functions, I decided to put a diagnostic in the entire file to fix it.
Do you agree ?

Sholde committed:
adding diagnostic in vprec vector file to avoid warning about changes abi

Sholde committed:
remove the assignement of useless pointer

Sholde committed:
separate libray flags to enable them only in linker stage

mistoan committed:
Fixes for the precision exploration using vfc_precexp (#247)

* modified libVFCFuncInstrument so that it now only uses mangled names

* bugfix for vfc_precexp, where in all mode one temporary variable was missing

* bugfix for handling run timeout in vfc_precexp

* bugfix for improper handling of the child process

* bugfix for vfc_precexp, where in all mode one temporary variable was missing

* bugfix for handling run timeout in vfc_precexp

* bugfix for improper handling of the child process

* fixed a broken link in the documentation

Co-authored-by: Matei Istoan <matei.istoan@inria.fr>

Sholde commented:
With my version, somebody who changes scalar functions must changes vector functions too in `vprec` and `ieee`.

titeup commented:
I just saw that you open this PR on your fork in safecarlo instead of verificarlo. Not sure if Pablo is following your work. At least if you prefer to do it there please invite him to join the PR/project.
But I think you can do the PR directly on verificarlo as it should be done to prepare merging in verificarlo master..

Sholde committed:
adding pre-configured setting in vprec backend (#249)

Fix #184.

Sholde committed:
clang format

Sholde committed:
clang format

Sholde committed:
adding support for gcc vector

Sholde committed:
adding gcc diagnostic

PurplePachyderm committed:
Integration of CI into master branch (#251)

* Integration of vfc_ci source code

Most of the source code/config files have already been modified.

For now, vfc_probes is added as a header (and not as a
precompiled library), which also required to add vfc_hashmap.h
as an accessible header. -> We might want to change this in the
future ?

sigdigits has also been added as a Git submodule, which is downloaded
in the "verificarlo" Makefile rule (there may be a better place to do
this, but it seemed to make sense since this action affects the entire
repository). It is installed as a Python submodule of verificarlo.

vfc_ci is added as a script, but not the ci folder: adding it to the
verificarlo Python does not work since the directory strucure (which is
needed, especially by Bokeh server) is flattened. We still need to
figure out a good way to copy the directory directly into pkgpython
with automake. Copying the folder manually seemed to work perfectly
though, so no other changes may be required to have a first functional
install setup.

* Copy ci directory into the verificarlo Python module

By using a separate Makefile.am for each ci subdirectory, and
overwriting the pkgpythondir variable, it is possible to copy all the source
files while keeping the same file structure.

* Move most Python scripts into a tools subdirectory

The src/tools subdirectory contains the Python scripts that were
previously in src and postprocessing.

All Makefiles and config files have been edited accordingly.

* Update vfc_ci with changes from dedicated repo

This includes support for multiple repository in one report and some
improvements to the code/comments.

* Replace vfc_put_probe by vfc_probe

* Remove sigdigits module

* Re-initialize sigdigits sumbodules in src/tools/sigdigits

* Add Python dependencies in Dockerfile and a vfc_ci test

Installing all the new Python dependencies in the Dockerfile should
solve the failing test issues, and result in a Docker image usable with
vfc_ci (including in Github/Gitlab CI).

To make sure that the install of vfc_ci is correct a new test has been
added : it tests the presence of the vfc_ci command itself, as well as
the vfc_probes system.

* Add documentation for Verificarlo CI

Also corrected a typo where "05-Postprocessing.md" was linked instead of
"06-Postprocessing.md", and another one with a wrong level of title.

* clang-format vfc_probes and corect permissions on vfc_ci test scripts

These 2 modifications should fix the clang-format-lint and test-docker
tests respectively

* Corrections from verificarlo/verificarlo#238

This includes improvements to the source code of vfc_ci itself
(optimizing some parts of the code, adiing license headers, ...) and to
the installation process (move submodules initialization, ...)

* Fix Dockerfile wit hPython dependencies

REmoved wrong apt packages. Moreover, some Python dependencies were
installed with both apt and pip : this has been removed to use apt if a
package is available, and pip otherwise.

* Custom timerange for run files in report & documentation correction

- It's now possible to specify if some of the most recent run files must
be ignored when generating the report. Coupling this option with a
custom number of files to show makes it possible to select any
chronological range of run files.

- Documentation has been corrected, it used to indicate that runb files
from the las 90 days would appear, instead of the last 100.

* Start adding the Fortran interface to Verificarlo

vfc_probes has been coverted to a shared object library, and the source
code of the Fortran interface has also been added.

Now, it needs to be added to the install process, only when the
--with-flang is specified.

* Update documentation with the Fortran interface

* Write complete test for Fortran interface

Even though the test itself should be correct, it is not expected to work in the current state of this branch, since it relies on vfc_probes_f.mod which still needs to be built wit automake. Moreover, it's unclear how Fortran code should be ignored/built depending on the --with-flang/--without-flang flags.

* Complete installation process ofr vfc_probes_f

The Fortran interface for vfc_probes is now installed when the
--with-flang is specified. If this is the case, the
test_fortran_vfc_probes will also be executed. The interface is
accessible with the vfc_probes_f library as well as its .mod file

* test.sh to test.sh.in in test_fortran_vfc_probes

Now the test.sh script will only be generated if Fortran support is
:x

* Fix Makefile issue in test_fortran_NAS

Co-authored-by: Aurélien Delval <aurelien@babtou-lent.babtou-lent>
Co-authored-by: Pablo de Oliveira Castro <pablo@sifflez.org>

Sholde committed:
supporting vector interface function only when verificarlo is configure with clang

Sholde committed:
remove warning of check vector interface functions

Sholde committed:
clang format lint

Sholde committed:
remove gcc vector types

Sholde committed:
adding gcc vector types

Sholde committed:
adding clang macro test

Sholde committed:
separate vector operations in cancellation

Sholde committed:
Merge branch 'master' into vectorization

Sholde committed:
update elements of the vector with the result value

Sholde committed:
Merge branch 'vectorization' of github.com:Sholde/verificarlo into vectorization

PurplePachyderm committed:
Makefile correction from verificarlo/verificarlo#253

Correctly ignores the rules for the Fortran interface when flang support
is disaled.

Also add a small correction to the code of the vfc_ci test command to
make it easier to correctly pass arguments to the test executables.

PurplePachyderm committed:
Merge pull request #255 from PurplePachyderm/master

Makefile correction from verificarlo/verificarlo#253

yohanchatelain committed:
Fix vector instrumentation + minor fixes: (#244)

* Fix vector instrumentation + minor fixes:
    - Vector instrumentation was invalid for unsupported vector:
      * New implementation used directly the signatures of vfcwrapper
      	in the instrumentation pass. To do so, we load the vfcwrapper
	IR file during the instrumentation pass.
      * Add test for checking vectorization instrumentation in
      	tests/test_vectorization
      * Check if ABI between the caller and the caller are identical and
        modify it if needed

    - Fix compilation of libVFCFuncInstrument.cpp for llvm-4.0
    - Update tests:
      * Remove test.sh.in generations for test_bitmask_backend,
      	test_fortran and test_fortran_NAS} and used paths.sh instead
      * Update clean.sh for some tests
      * Update test_branch_instrumentation test to check if scalar and
      	vector instrumentation give the same result
      * Fix test_cpp when using llvm-4
    - Fix -warnings

yohanchatelain committed:
Fix vfc_probes error when compiling with Fortran: (#258)

* Fix vfc_probes error when compiling with Fortran:
    - Fix module incompatibility when we compile libvfc_probes_f.a
      with gfortran and compiling codes with flang
    - Force flang to be used to compile libvfc_probes_f
    - Pass include path during compilation in verificarlo.in.in
    - Make src/tools/ci/test_data_processing.py retro compatible
      with older scipy version that don't use namedtuple for returned
      value

Sholde committed:
fix the removing of init and exit functions in vprec

Sholde committed:
Merge branch 'master' into vectorization
